### DISCUSSION TITLE
How to communicate with hacktribe via SysEx.

### DISCUSSION THREAD

**Discussion by bangcorrupt** [bangcorrupt/hacktribe#37] [Q&A] [answered]
#30 @gradientledge asked:
> so for the sysex are you just using midiox to communicate with it, or is there a better program as that ones ancient?

**Comment by bangcorrupt** [comment, marked as answer]
I use python and [mido](https://mido.readthedocs.io/en/latest/).  In the [hacktribe-editor](https://github.com/bangcorrupt/hacktribe-editor) repo there is [e2sysex.py](https://github.com/bangcorrupt/hacktribe-editor/blob/main/e2sysex.py) with functions for getting and setting patterns.  As I add new SysEx features to hacktribe I will add functions to that script.  

I can't give helpful advice on Windows, I haven't used it in over a decade.

**Comment by bangcorrupt** [comment]
#22 @ShoveD01 asked:
> what are the sysex strings that currently work with the e2?

All SysEx messages defined in [Electribe 2 Sampler MIDI Implementation](https://www.korg.com/us/support/download/manual/0/368/2665/).

I've added SysEx functions for arbitrary read/write of CPU RAM and flash.  The best way to use these is via [e2sysex.py](https://github.com/bangcorrupt/e2-scripts/blob/master/e2sysex.py).  This includes helper functions to get and set data such as groove templates and IFX presets.  Writing these to the flash memory will be added in a later update.

**Comment by chaocrator** [comment]
on Linux i'm using sendmidi. 
available for other platforms as well:
https://github.com/gbevin/SendMIDI/releases

**Reply by rafamj** [reply]
This program uses Electribe sysex: https://github.com/rafamj/elecmidi

**Comment by maks** [comment]
So I've started reading through that sampler midi implementation doc from Korg for a project Im working on and I've run into a few interesting things - I wonder if @bangcorrupt or @rafamj either of you may know about these?

First off, the sampler doc is almost identical to the synth one, so pattern data for each part only has a field for "Oscillator Type":
`| 8~9         | Oscillator Type          | 0~500         `
so I can't figure out how do patterns specify which user samples (>500) are used for a part? I haven't actually had a chance to test this out, so I wonder if that doc is just wrong and those 2 bytes can actually store sample slots by having a count >500 ?

Also I can't see anywhere in the Global data dump fields where the oscillator/sample names for each slot number are - I guess Korg didnt bother to include that so @bangcorrupt I wonder if that something you may have access to and perhaps could add as a future sysex request-response, somewhat like you did for IFX?

**Reply by bangcorrupt** [reply]
I haven't checked but I think you're right about the doc being wrong.  

The Global Settings sysex dump includes only the things in the Global Settings menu. I can add functions to get the oscillator presets at some point.  The new sysex functions I've added all use an arbitrary read/wriite function in the firmware.  Selecting specific data and extracting parameters is all done in python.  We can use `read_cpu_ram(address, length)` to get data from any address the CPU can access.

**Reply by maks** [reply]
Awesome work @bangcorrupt ! I think `read_cpu_ram()` exactly what I need! 👍🏻  and double thanks for the starting address, I'll POKE around in there and see if I can pull out the strings I want.

**Reply by rafamj** [reply]
Hello,

  I have the synth Electribe. When sent a value from 0 to 408 the osc changes accordingly form 1 to 409. For values over 408 the osc type seemingly changes to <value>%408 but the oscilator name displayed changes randomly.

I have modified elecmidi to include the parameter oscType. For example one file with the 3 following lines would send the value 217 to oscType of part 1 and change the type to Darbuka1 
read
oscType 1 217
write

**Reply by maks** [reply]
Reporting back to follow up on my own question, it turns out that the docs are just wrong and that user samples in >500 slots are stored in the oscType just fine on the hacktribe firmware.

**Comment by bangcorrupt** [comment]
If you get buffer overruns or error messages when trying to send pattern files, try installing https://github.com/bangcorrupt/mido and https://github.com/bangcorrupt/python-rtmidi.  These will split long sysex messages into 512 byte chunks.  Sending other MIDI messages while a split sysex message is transmitting can break transmission.

**Reply by rafamj** [reply]
Many thanks. 

  I have incorporated this idea to my program, sending 512 byte chunks.   It runs without errors when the electribe is stopped, but when it is playing a pattern I still get some errors.

**Reply by bangcorrupt** [reply]
Thank you for testing. What errors do you get, are they from the OS or the electribe?  It may just need smaller chunks, or a longer delay between them.

**Comment by rafamj** [comment]
I have only put a delay every 512 bytes in my program. It seems that when the electribe is running it has no power to process de data.
